File: src/modules/flatpak_manager.rs
use crate::get_image_type;
use colored::Colorize;
use std::fs;
use std::path::Path;
use std::path::PathBuf;
use std::process::Command;

fn remove_fedora_remote() {
    let mut remove_command = Command::new("flatpak")
        .args(["remote-delete", "--force", "fedora"])
        .spawn()
        .expect("Failed to start flatpak remote-delete");

    let status = remove_command.wait().expect("Failed to remove the remote");

    if !status.success() {
        eprintln!(
            "{}: Failed to remove the Fedora remote!",
            "ERR".bold().red()
        )
    }
}

fn remove_flatpaks() {
    let mut flatpaks_to_remove: Vec<PathBuf> = Vec::new();
    let image_type = get_image_type();

    match image_type {
        _ if image_type.contains("aster") => {
            flatpaks_to_remove.push(PathBuf::from(
                "/usr/share/horizon/packages/aster/aster_flatpak_remove",
            ));
        }

        _ if image_type.contains("arcturus") => return,

        _ if image_type.contains("nova") => return,

        _ => {
            eprintln!("No matching image type for flatpak installation");
            return;
        }
    }

    for flatpak_list in flatpaks_to_remove {
        let contents = fs::read_to_string(flatpak_list).expect("Failed to read flatpaks list");

        for line in contents.lines() {
            if !line.trim().is_empty() {
                let status = Command::new("flatpak")
                    .args(["remove", "--system", "-y", line.trim()])
                    .status()
                    .expect("Failed to run flatpak remove");

                if !status.success() {
                    eprintln!("Failed to remove {line}");
                }
            }
        }
    }
}

pub fn fix_theming() {
    // Helper function to run flatpak override commands
    fn run_flatpak_override(args: &[&str]) {
        let status = Command::new("flatpak")
            .arg("override")
            .args(args)
            .status()
            .expect("Failed to run flatpak override");

        if status.success() {
            println!("{}: flatpak override {:?}", "DONE".bold().green(), args);
        } else {
            eprintln!("{}: flatpak override {:?} failed", "ERR".bold().red(), args);
        }
    }

    // Flatpak theming support
    run_flatpak_override(&[
        "--filesystem=xdg-config/gtk-4.0:ro",
        "--filesystem=xdg-config/gtk-3.0:ro",
        "--filesystem=xdg-data/icons:ro",
    ]);

    // XInput for Firefox
    run_flatpak_override(&["--system", "--env=MOZ_USE_XINPUT2=1", "org.mozilla.firefox"]);

    // Fix printing on LibreOffice
    run_flatpak_override(&[
        "--system",
        "--socket=cups",
        "--socket=session-bus",
        "org.libreoffice.LibreOffice",
    ]);

    // Allow MangoHUD config for Flatpaks
    run_flatpak_override(&[
        "--filesystem=xdg-config/MangoHud:ro",
        "--filesystem=xdg-config/vkBasalt:ro",
    ]);

    // Fix permissions for XIV Launcher
    run_flatpak_override(&["--device=dri", "dev.goats.xivlauncher"]);

    // Fix permissions for Protontricks
    run_flatpak_override(&[
        "--filesystem=~/.local/share/Steam",
        "--filesystem=/var/mnt",
        "--filesystem=/run/media",
        "com.github.Matoking.protontricks",
    ]);

    // Nvidia-specific logic using get_image_type()
    let image_type = get_image_type(); // Uses your existing logic

    if image_type.contains("nvidia") {
        let lshw_output = Command::new("lshw")
            .arg("-C")
            .arg("display")
            .output()
            .expect("Failed to run lshw");

        let lshw_stdout = String::from_utf8_lossy(&lshw_output.stdout);

        let nvidia_present = lshw_stdout.contains("vendor: NVIDIA Corporation");
        let display_count = lshw_stdout.matches("-display").count();

        if nvidia_present && display_count <= 1 {
            // Apply Nvidia-specific flatpak override
            run_flatpak_override(&[
                "--system",
                "--filesystem=host-os",
                "--env=LIBVA_DRIVER_NAME=nvidia",
                "--env=LIBVA_DRIVERS_PATH=/run/host/usr/lib64/dri",
                "--env=LIBVA_MESSAGING_LEVEL=1",
                "--env=MOZ_DISABLE_RDD_SANDBOX=1",
                "--env=NVD_BACKEND=direct",
                "org.mozilla.firefox",
            ]);
        } else {
            // Undo Nvidia-specific overrides
            run_flatpak_override(&[
                "--system",
                "--nofilesystem=host-os",
                "--unset-env=LIBVA_DRIVER_NAME",
                "--unset-env=LIBVA_DRIVERS_PATH",
                "--unset-env=LIBVA_MESSAGING_LEVEL",
                "--unset-env=MOZ_DISABLE_RDD_SANDBOX",
                "--unset-env=NVD_BACKEND",
                "org.mozilla.firefox",
            ]);
        }
    } else {
        // Non-NVIDIA image: Ensure overrides are cleared
        run_flatpak_override(&[
            "--system",
            "--nofilesystem=host-os",
            "--unset-env=LIBVA_DRIVER_NAME",
            "--unset-env=LIBVA_DRIVERS_PATH",
            "--unset-env=LIBVA_MESSAGING_LEVEL",
            "--unset-env=MOZ_DISABLE_RDD_SANDBOX",
            "--unset-env=NVD_BACKEND",
            "org.mozilla.firefox",
        ]);
    }
}

fn enable_flathub() {
    println!("");

    let mut add_command = Command::new("flatpak")
        .args([
            "remote-add",
            "--if-not-exists",
            "--system",
            "flathub",
            "https://dl.flathub.org/repo/flathub.flatpakrepo",
        ])
        .spawn()
        .expect("Failed to start flatpak remote-add");

    let status = add_command.wait().expect("Failed to add flathub");

    if !status.success() {
        eprintln!("{}: Failed to add flathub!", "ERR".bold().red())
    }

    let mut enable_command = Command::new("flatpak")
        .args(["remote-modify", "--system", "--enable", "flathub"])
        .spawn()
        .expect("Failed to start flatpak remote-modify");

    let status = enable_command.wait().expect("Failed to modify flathub");

    if !status.success() {
        eprintln!("{}: Failed to modify flathub!", "ERR".bold().red())
    }
}

fn install_flatpaks() {
    let mut flatpaks_to_install: Vec<PathBuf> = Vec::new();
    let image_type = get_image_type();

    match image_type {
        _ if image_type.contains("aster") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/horizon/packages/aster/aster_flatpaks",
            ));

            flatpaks_to_install.push(PathBuf::from("/usr/share/horizon/packages/shared_flatpaks"));
        }

        // Annoying thing
        _ if image_type.contains("arcturus") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/horizon/packages/arcturus_flatpaks",
            ));

            flatpaks_to_install.push(PathBuf::from("/usr/share/horizon/packages/shared_flatpaks"));
        }

        _ if image_type.contains("umbra") => {
            flatpaks_to_install.push(PathBuf::from("/usr/share/umbra/packages/flatpaks"));
        }
        _ if image_type.contains("nova_gnome") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/gnome/install_flatpaks",
            ));
        }
        _ if image_type.contains("nova_gnome_dx") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/gnome/install_flatpaks_dx",
            ));
        }
        _ if image_type.contains("nova_gnome_gaming") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/gnome/install_flatpaks_gaming",
            ));
        }
        _ if image_type.contains("supernova_gnome") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/gnome/install_flatpaks",
            ));
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/gnome/install_flatpaks_dx",
            ));
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/gnome/install_flatpaks_gaming",
            ));
        }
        _ if image_type.contains("nova_plasma") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/plasma/install_flatpaks",
            ));
        }
        _ if image_type.contains("nova_plasma_dx") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/plasma/install_flatpaks_dx",
            ));
        }
        _ if image_type.contains("nova_plasma_gaming") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/plasma/install_flatpaks_gaming",
            ));
        }
        _ if image_type.contains("supernova_plasma") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/plasma/install_flatpaks",
            ));
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/plasma/install_flatpaks_dx",
            ));
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/plasma/install_flatpaks_gaming",
            ));
        }
        _ if image_type.contains("nova_cosmic_dx") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/cosmic/install_flatpaks_dx",
            ));
        }
        _ if image_type.contains("supernova_cosmic") => {
            flatpaks_to_install.push(PathBuf::from(
                "/usr/share/nova/packages/cosmic/install_flatpaks_dx",
            ));
        }

        _ => {
            eprintln!("No matching image type for flatpak installation");
            return;
        }
    }

    for flatpak_list in flatpaks_to_install {
        let contents = fs::read_to_string(flatpak_list).expect("Failed to read flatpaks list");

        for line in contents.lines() {
            if !line.trim().is_empty() {
                let status = Command::new("flatpak")
                    .args(["install", "--system", "-y", line.trim()])
                    .status()
                    .expect("Failed to run flatpak install");

                if !status.success() {
                    eprintln!("Failed to install {line}");
                }
            }
        }
    }
}

fn disable_systemd_service() {
    let _systemd_service_path: Option<&Path>;
    let service_name: &str;

    // Determine which systemd service exists
    if Path::new("/usr/share/horizon").exists() {
        _systemd_service_path = Some(Path::new(
            "/usr/lib/systemd/system/horizon-flatpak-manager.service",
        ));
        service_name = "horizon-flatpak-manager.service";
    } else if Path::new("/usr/share/nova").exists() {
        _systemd_service_path = Some(Path::new(
            "/usr/lib/systemd/system/nova-flatpak-manager.service",
        ));
        service_name = "nova-flatpak-manager.service";
    } else if Path::new("/usr/share/umbra").exists() {
        _systemd_service_path = Some(Path::new(
            "/usr/lib/systemd/system/umbra-flatpak-manager.service",
        ));
        service_name = "umbra-flatpak-manager.service";
    } else {
        panic!("{}: No systemd service found!", "ERR".bold().red());
    };

    // Disable the systemd service using systemctl
    let status = Command::new("systemctl")
        .args(["disable", "--now", service_name])
        .status()
        .expect("Failed to run systemctl disable command");

    if status.success() {
        println!(
            "{}: Successfully disabled {}",
            "DONE".bold().green(),
            service_name
        );
    } else {
        eprintln!("{}: Failed to disable {}", "ERR".bold().red(), service_name);
    }
}

pub fn flatpak_manager_remove() {
    remove_flatpaks();
}

pub fn flatpak_manager() {
    remove_fedora_remote();
    enable_flathub();
    install_flatpaks();
    fix_theming();
    disable_systemd_service();
}

File: modules/flatpakManager.go
package modules

import (
	"bytes"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/darkmatterorg/orbit/utils"
)

var image_type = utils.Getimagetype()

func check_fedora_remote() bool {
	cmd := exec.Command("flatpak", "remotes")
	var out bytes.Buffer
	cmd.Stdout = &out
	err := cmd.Run()

	if err != nil {
		utils.Error("Failed to run command.")
	}

	output := out.String()
	if strings.Contains(output, "fedora") {
		return true
	}

	return false
}

func remove_fedora_remote() {
	if check_fedora_remote() {
		return
	}

	cmd := exec.Command("flatpak", "remote-delete", "--force", "fedora")
	err := cmd.Run()

	if err != nil {
		utils.Error("Failed to remove the fedora remote.")
	}
}

func remove_flatpaks() {
	var flatpaksToRemove []string

	switch {
	case strings.Contains(image_type, "aster"):
		flatpaksToRemove = append(flatpaksToRemove, filepath.Join("/usr/share/horizon/packages/aster/aster_flatpak_remove"))
	case strings.Contains(image_type, "arcturus"), strings.Contains(image_type, "nova"):
		return
	default:
		utils.Notice("Not implemented yet.")
		return
	}

	for _, flatpakList := range flatpaksToRemove {
		contents, err := os.ReadFile(flatpakList)
		if err != nil {
			var error_message = fmt.Sprintf("Full Error: %v\n", err)

			utils.Error("Failed to read the flatpaks list.\n" + error_message)
			continue
		}

		lines := strings.Split(string(contents), "\n")
		for _, line := range lines {
			line = strings.TrimSpace(line)
			if line != "" {
				cmd := exec.Command("flatpak", "remove", "--system", "-y", line)
				_, err := cmd.CombinedOutput()
				if err != nil {
					utils.Error("Failed to remove flatpaks from list.")
				} else {
					utils.Done("Removed all flatpaks from list.")
				}
			}
		}
	}
}

func run_flatpak_override(args []string) {
	cmd := exec.Command("flatpak", append([]string{"override"}, args...)...)
	err := cmd.Run()

	if err != nil {
		var message = fmt.Sprintf("flatpak override " + strings.Join(args, " "))
		utils.Done(message)
	}
}

func fix_theming() {
	run_flatpak_override([]string{
		"--filesystem=xdg-config/gtk-4.0:ro",
		"--filesystem=xdg-config/gtk-3.0:ro",
		"--filesystem=xdg-data/icons:ro",
	})

	// XInput for Firefox
	run_flatpak_override([]string{"--system", "--env=MOZ_USE_XINPUT2=1", "org.mozilla.firefox"})

	// Fix printing on LibreOffice
	run_flatpak_override([]string{
		"--system",
		"--socket=cups",
		"--socket=session-bus",
		"org.libreoffice.LibreOffice",
	})

	// Allow MangoHUD config for Flatpaks
	run_flatpak_override([]string{
		"--filesystem=xdg-config/MangoHud:ro",
		"--filesystem=xdg-config/vkBasalt:ro",
	})

	// Fix permissions for XIV Launcher
	run_flatpak_override([]string{"--device=dri", "dev.goats.xivlauncher"})

	// Fix permissions for Protontricks
	run_flatpak_override([]string{
		"--filesystem=~/.local/share/Steam",
		"--filesystem=/var/mnt",
		"--filesystem=/run/media",
		"com.github.Matoking.protontricks",
	})

	// Nvidia-specific logic using getImageType()

	if strings.Contains(image_type, "nvidia") {
		lshwOutput, err := exec.Command("lshw", "-C", "display").Output()
		if err != nil {
			utils.Error("Failed to run lshw")
			return
		}

		lshwStdout := string(lshwOutput)
		nvidiaPresent := strings.Contains(lshwStdout, "vendor: NVIDIA Corporation")
		displayCount := strings.Count(lshwStdout, "-display")

		if nvidiaPresent && displayCount <= 1 {
			// Apply Nvidia-specific flatpak override
			run_flatpak_override([]string{
				"--system",
				"--filesystem=host-os",
				"--env=LIBVA_DRIVER_NAME=nvidia",
				"--env=LIBVA_DRIVERS_PATH=/run/host/usr/lib64/dri",
				"--env=LIBVA_MESSAGING_LEVEL=1",
				"--env=MOZ_DISABLE_RDD_SANDBOX=1",
				"--env=NVD_BACKEND=direct",
				"org.mozilla.firefox",
			})
		} else {
			// Undo Nvidia-specific overrides
			run_flatpak_override([]string{
				"--system",
				"--nofilesystem=host-os",
				"--unset-env=LIBVA_DRIVER_NAME",
				"--unset-env=LIBVA_DRIVERS_PATH",
				"--unset-env=LIBVA_MESSAGING_LEVEL",
				"--unset-env=MOZ_DISABLE_RDD_SANDBOX",
				"--unset-env=NVD_BACKEND",
				"org.mozilla.firefox",
			})
		}
	} else {
		// Non-NVIDIA image: Ensure overrides are cleared
		run_flatpak_override([]string{
			"--system",
			"--nofilesystem=host-os",
			"--unset-env=LIBVA_DRIVER_NAME",
			"--unset-env=LIBVA_DRIVERS_PATH",
			"--unset-env=LIBVA_MESSAGING_LEVEL",
			"--unset-env=MOZ_DISABLE_RDD_SANDBOX",
			"--unset-env=NVD_BACKEND",
			"org.mozilla.firefox",
		})
	}
}

func enable_flathub() {
	// Add flathub
	add_flathub := exec.Command("flatpak", "remote-add", "--if-not-exists", "--system", "flathub", "https://dl.flathub.org/repo/flathub.flatpakrepo")
	add_flathub_err := add_flathub.Run()

	if add_flathub_err != nil {
		utils.Error("Failed to add flathub")
	}

	// Enable flathub
	enable_flathub := exec.Command("flatpak", "remote-modify", "--system", "--enable", "flathub")
	enable_flathub_err := enable_flathub.Run()

	if enable_flathub_err != nil {
		utils.Error("Failed to enable flathub.")
	}

}

func install_flatpaks() {
	var flatpaks_to_install []string

	switch {
	case strings.Contains(image_type, "aster"):
		flatpaks_to_install = append(flatpaks_to_install, filepath.Join("/usr/share/horizon/packages/aster/aster_flatpaks"))
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/horizon/packages/shared_flatpaks")

	case strings.Contains(image_type, "arcturus"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/horizon/packages/arcturus_flatpaks")
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/horizon/packages/shared_flatpaks")

	case strings.Contains(image_type, "umbra"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/umbra/packages/flatpaks")

	case strings.Contains(image_type, "nova_gnome"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/gnome/install_flatpaks")

	case strings.Contains(image_type, "nova_gnome_dx"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/gnome/install_flatpaks_dx")

	case strings.Contains(image_type, "nova_gnome_gaming"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/gnome/install_flatpaks_gaming")

	case strings.Contains(image_type, "supernova_gnome"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/gnome/install_flatpaks")
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/gnome/install_flatpaks_dx")
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/gnome/install_flatpaks_gaming")

	case strings.Contains(image_type, "nova_plasma"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/plasma/install_flatpaks")

	case strings.Contains(image_type, "nova_plasma_dx"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/plasma/install_flatpaks_dx")

	case strings.Contains(image_type, "nova_plasma_gaming"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/plasma/install_flatpaks_gaming")

	case strings.Contains(image_type, "supernova_plasma"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/plasma/install_flatpaks")
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/plasma/install_flatpaks_dx")
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/plasma/install_flatpaks_gaming")

	case strings.Contains(image_type, "nova_cosmic_dx"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/cosmic/install_flatpaks_dx")

	case strings.Contains(image_type, "supernova_cosmic"):
		flatpaks_to_install = append(flatpaks_to_install, "/usr/share/nova/packages/cosmic/install_flatpaks_dx")
	default:
		utils.Notice("Not implemented yet.")
		return
	}

	for _, flatpakList := range flatpaks_to_install {
		contents, err := os.ReadFile(flatpakList)
		if err != nil {
			var error_message = fmt.Sprintf("Full Error: %v\n", err)

			utils.Error("Failed to read the flatpaks list.\n" + error_message)
			continue
		}

		lines := strings.Split(string(contents), "\n")
		for _, line := range lines {
			line = strings.TrimSpace(line)
			if line != "" {
				cmd := exec.Command("flatpak", "remove", "--system", "-y", line)
				_, err := cmd.CombinedOutput()
				if err != nil {
					utils.Error("Failed to install flatpaks from list.")
				} else {
					utils.Done("Installed all flatpaks from list.")
				}
			}
		}
	}
}

func Flatpak_manager_remove() {
	remove_flatpaks()

	switch {
	case strings.Contains(image_type, "nova"):
		utils.Disable_systemd_service("nova-flatpak-manager")
	case strings.Contains(image_type, "arcturus"), strings.Contains(image_type, "aster"):
		utils.Disable_systemd_service("horizon-flatpak-manager")
	case strings.Contains(image_type, "umbra"):
		utils.Disable_systemd_service("umbra-flatpak-manager")
	default:
		utils.Error("Image not supported.")
	}
}

func Flatpak_manager() {
	remove_fedora_remote()
	enable_flathub()
	install_flatpaks()
	fix_theming()

	switch {
	case strings.Contains(image_type, "nova"):
		utils.Disable_systemd_service("nova-flatpak-manager")
	case strings.Contains(image_type, "arcturus"), strings.Contains(image_type, "aster"):
		utils.Disable_systemd_service("horizon-flatpak-manager")
	case strings.Contains(image_type, "umbra"):
		utils.Disable_systemd_service("umbra-flatpak-manager")
	default:
		utils.Error("Image not supported.")
	}

}
